### PR TITLE
all: preallocate slices where possible

### PR DESCRIPTION
**Pull request by amenzhinsky**
I'd spotted few places in xds client code where slices could be preallocated to avoid extra allocations, but then I simply used https://github.com/alexkohler/prealloc to track down similar issues throughout the codebase.

RELEASE NOTES: N/A

### CODE REVIEW

File: balancer/grpclb/grpclb_remote_balancer.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ func (lb *lbBalancer) processServerList(l *lbpb.ServerList) {
 	}
 	lb.fullServerList = l.Servers
 
-	var backendAddrs []resolver.Address
+	backendAddrs := make([]resolver.Address, 0, len(l.Servers))
```
Comment (easwars) [inline balancer/grpclb/grpclb_remote_balancer.go:66]: There are cases where we end up *not* adding an entry to the slice from within the `for` loop and instead `continue`. So, it would be better to keep this as is.

File: internal/resolver/dns/dns_resolver_test.go [hunk 2]
```diff
@@ -714,17 +714,17 @@ func testDNSResolver(t *testing.T) {
...
 		{
 			"srv.ipv6.single.fake",
-			nil,
+			[]resolver.Address{},
 			generateSC("srv.ipv6.single.fake"),
 		},
 		{
 			"srv.ipv6.multi.fake",
-			nil,
+			[]resolver.Address{},
```
Comment (easwars) [inline internal/resolver/dns/dns_resolver_test.go:727]: Why is this change required? We should probably leave it as is.

File: xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go [hunk 3]
```diff
@@ -176,6 +170,9 @@ func buildClusterImplConfigForEDS(parentPriority int, edsResp xdsclient.Endpoint
 	}
 
 	priorityChildNames, priorities := groupLocalitiesByPriority(edsResp.Localities)
+	retNames := make([]string, 0, len(priorityChildNames))
+	retAddrs := make([]resolver.Address, 0, len(priorityChildNames))
```
Comment (easwars) [inline xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go:174]: We could get more than one entry added to `retAddrs` per iteration of the `for` loop. So, we should revert this change.

File: xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go [hunk 4]
```diff
@@ -205,8 +202,8 @@ func buildClusterImplConfigForEDS(parentPriority int, edsResp xdsclient.Endpoint
 // - ["p0", "p1"]
 // - map{"p0":[L0, L1], "p1":[L2]}
 func groupLocalitiesByPriority(localities []xdsclient.Locality) ([]string, map[string][]xdsclient.Locality) {
-	var priorityIntSlice []int
-	priorities := make(map[string][]xdsclient.Locality)
+	priorities := make(map[string][]xdsclient.Locality, len(localities))
+	priorityIntSlice := make([]int, 0, len(localities))
```
Comment (easwars) [inline xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go:206]: Same here. There is a `continue` within the `for` loop. So, we should revert this.

File: xds/internal/xdsclient/filter_chain.go [hunk 5]
```diff
@@ -500,7 +503,7 @@ func filterByDestinationPrefixes(dstPrefixes []*destPrefixEntry, isUnspecified b
 		return dstPrefixes
 	}
 
-	var matchingDstPrefixes []*destPrefixEntry
+	matchingDstPrefixes := make([]*destPrefixEntry, 0, len(dstPrefixes))
```
Comment (easwars) [inline xds/internal/xdsclient/filter_chain.go:506]: We have multiple `continue`s within the `for` loop. We should revert this.

File: xds/internal/xdsclient/rds_test.go [hunk 6]
```diff
@@ -1378,6 +1378,7 @@ func (s) TestHashPoliciesProtoToSlice(t *testing.T) {
 			hashPolicies: []*v3routepb.RouteAction_HashPolicy{
 				{PolicySpecifier: &v3routepb.RouteAction_HashPolicy_FilterState_{FilterState: &v3routepb.RouteAction_HashPolicy_FilterState{Key: "unsupported key"}}},
 			},
+			wantHashPolicies: []*HashPolicy{},
```
Comment (easwars) [inline xds/internal/xdsclient/rds_test.go:1381]: This is not required.

File: xds/internal/xdsclient/xds.go [hunk 7]
```diff
@@ -406,8 +406,7 @@ func generateRDSUpdateFromRouteConfiguration(rc *v3routepb.RouteConfiguration, l
 }
 
 func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger, v2 bool) ([]*Route, error) {
-	var routesRet []*Route
-
+	routesRet := make([]*Route, 0, len(routes))
```
Comment (easwars) [inline xds/internal/xdsclient/xds.go:409]: We have a `continue` within the `for` loop. This should be reverted.

**Review by easwars — CHANGES_REQUESTED** [review]

File: xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go (hunk 3 above)
Comment (amenzhinsky) [reply xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go:174]: In this case we're allocating at least the minimal capacity, I think it's better than starting from zero, because it grows exponentially 1,2,4,8 it does helps with small `priorityChildNames` numbers.

File: internal/resolver/dns/dns_resolver_test.go (hunk 2 above)
Comment (amenzhinsky) [reply internal/resolver/dns/dns_resolver_test.go:727]: The fake host resolver that's used in the tests returns empty slices for `*.fake` addresses and the test ends up comparing `reflect.DeepEqual([]string{}, nil)` that fails.

`net.LookupHost` guarantees that slice is not empty when error is nil but the test expects `state` to be updated with an empty list of addresses.

Comment (easwars) [reply internal/resolver/dns/dns_resolver_test.go:727]: I think we should keep this as `nil`, and change the `reflect.DeepEqual` to `cmp.Equal` to which you can pass a `cmpopts.EquateEmpty` which is meant exactly for this usecase.

File: xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go (hunk 3 above)
Comment (easwars) [reply xds/internal/balancer/clusterresolver/balancerconfig/configbuilder.go:174]: Ack.

**Comment by easwars** [comment]
Could you please rebase to resolve the conflicts.

Also, if you can avoid force pushing whenever possible, that would be great. GitHub loses track of comments when a force push is done. Thanks.

**Review by easwars — APPROVED** [review]
Thank you very much for the PR !!
